### ISSUE TITLE
Confusion about the implementation for generating the first token

### ISSUE THREAD

**Issue by oximi123** [facebookresearch/llama#674]
Hey guys, I am confused about the following code at generation.py, starts at line 136:
`        for cur_pos in range(min_prompt_len, total_len):
            logits = self.model.forward(tokens[:, prev_pos:cur_pos], prev_pos)
            if logprobs:
                token_logprobs[:, prev_pos + 1 : cur_pos + 1] = -F.cross_entropy(
                    input=logits.transpose(1, 2),
                    target=tokens[:, prev_pos + 1 : cur_pos + 1],
                    reduction="none",
                    ignore_index=pad_id,
                )`
 Here we see the **cur_pos** starts at the **min_prompt_len**, It actually means that the prompt with the **min_prompt_len** will be processed differently (i.e., the attention mechnism) from other prompt at the first iteration. Specifically, it takes all the promts to compute the first token for the prompt with the **min_prompt_len**, while for other prompts it only takes the previous one token for computing the frist token. I wonder if it is ok to do so or just a trick (I also wonder if other large language model use this trick, too)

**Comment by anthony-wss** [comment]
Same question here. Does anyone have any ideas?

**Comment by coming0228** [comment]
The keys, values embeddings in the Attention module that were already computed in the past are stored in the cache ([code](https://github.com/facebookresearch/llama/blob/main/llama/model.py#L285-L289)). So effectively, all the prompts with length greater than **min_promp_len** still uses all the previous prompt token embeddings (rather than just previous one token) to compute the first token in the attention module. The cache mechanism avoids re-computing the same thing over and over again and is good optimization trick.

**Comment by coming0228** [comment]
Close the issue. Please feel free to comment if you have further questions.
